portal.go

```diff
@@ -2047,22 +2063,65 @@ func (portal *Portal) HandleiMessage(msg *imessage.Message) id.EventID {
 		portal.bridge.IM.SendMessageBridgeResult(msg.ChatGUID, msg.GUID, eventID, overrideSuccess || hasMXID)
 	}()
 
+	// Look up the message in the database
+	dbMessage = portal.bridge.DB.Message.GetLastByGUID(portal.GUID, msg.GUID)
+
 	if portal.IsPrivateChat() && msg.ChatGUID != portal.LastSeenHandle {
 		portal.log.Debugfln("Updating last seen handle from %s to %s", portal.LastSeenHandle, msg.ChatGUID)
 		portal.LastSeenHandle = msg.ChatGUID
 		portal.Update(nil)
 	}
 
+	// Handle message tapbacks
 	if msg.Tapback != nil {
 		portal.HandleiMessageTapback(msg)
 		return ""
-	} else if portal.bridge.DB.Message.GetLastByGUID(portal.GUID, msg.GUID) != nil {
-		portal.log.Debugln("Ignoring duplicate message", msg.GUID)
-		// Send a success confirmation since it's a duplicate message
-		overrideSuccess = true
+	}
+
+	// If the message exists in the database, handle edits or retractions
+	if dbMessage != nil && dbMessage.MXID != "" {
+
+		// Clean up the text so the length is accurate
+		msg.Text = strings.ReplaceAll(msg.Text, "\ufffc", "")
+		msg.Subject = strings.ReplaceAll(msg.Subject, "\ufffc", "")
+
+		// DEVNOTE: It seems sometimes the message is just edited to remove data instead of actually retracting it
+		if msg.IsRetracted ||
+			(len(msg.Attachments) == 0 && len(msg.Text) == 0 && len(msg.Subject) == 0) {
+
+			// Retract existing message
+			if portal.HandleMessageRevoke(*msg) {
+				portal.zlog.Debug().Str("messageGUID", msg.GUID).Str("chatGUID", msg.ChatGUID).Msg("Revoked message")
+			} else {
+				portal.zlog.Warn().Str("messageGUID", msg.GUID).Str("chatGUID", msg.ChatGUID).Msg("Failed to revoke message")
+			}
+
+			overrideSuccess = true
+		} else if msg.IsEdited {
+
+			// Edit existing message
+			intent := portal.getIntentForMessage(msg, nil)
+			portal.handleNormaliMessage(msg, dbMessage, intent, &dbMessage.MXID)
+
+			overrideSuccess = true
+		} else if msg.IsRead && msg.IsFromMe {
+
+			// Send read receipt
+			err := portal.markRead(portal.MainIntent(), dbMessage.MXID, msg.ReadAt)
+			if err != nil {
+				portal.log.Warnfln("Failed to send read receipt for %s: %v", dbMessage.MXID, err)
+			}
+
+			overrideSuccess = true
```

Comment:
Why is this here 🤔

Reply:
BlueBubbles doesn't actually send a specific event for read receipts, just a generic `updated-message` event which could be an unsend, edit, delivery receipt, or read receipt. Given that you [already have code](https://github.com/mautrix/imessage/blob/6118c551ac9d840e6ca2e5ab1c30efc8f60549c3/portal.go#L2108) in this same function which sends a read receipt for a new message, I felt it was appropriate to interpret if this event is actually a message being read. Otherwise I would have to do all the previous checks this function just did, but in the bluebubbles code.